torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+    return;
+  }
+
+  // Search for vertices that are reachable from consumer that have a now incorrect
+  // topological ordering.
+  if (!dfsForward(consumer, producer->ord, deltaF_)) {
+    // dfsForward returns false if there is a path from consumer to producer.
+    throw std::runtime_error("Cycle detected while trying to add edge.");
+  }
+
+  // Search for vertices that can reach producer that have a now incorrect
+  // topological ordering
+  dfsBackward(producer, consumer->ord, deltaB_);
+
+  // Reorder the vertices that are reachable from consumer to occur BEFORE
+  // the vertices that can reach producer.
+  reorder(deltaF_, deltaB_);
+}
+
+// Define the affected region for contractEdge(producer, consumer) as
+// { v s.t. ord(v) in [ord(producer), ord(consumer)] }.
+// These are the only vertices that can possibly be moved around
+// during edge contraction.
+//
+// contractEdge is O(|AR| * |out_edges(producer)|)
+template <typename T>
+bool DynamicDAG<T>::contractEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (contractionProducesCycle(producer, consumer)) {
+    return false;
+  }
+
+  removeEdge(producer, consumer);
+
+  // We can either merge producer into consumer or the other way around.
+  // I've chosen to merge producer into consumer for now but this is arbitrary.
+  consumer->rdata.insert(consumer->rdata.end(), producer->rdata.begin(), producer->rdata.end());
+
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+  std::tie(in_edges, out_edges) = removeVertex(producer);
+
+  // Each of these are constant b/c ord(consumer) > ord(producer) > ord(parent)
+  // so the edge addition still preserves the existing topological order.
+  for (auto* parent : in_edges) {
+    addEdge(parent, consumer);
+  }
+
+  // NB: each addEdge call is linear in (ord(consumer) - ord(child)).
+  // This makes contractEdges O(|out_edges(producer)| * |AR|).
+  for (auto* child : out_edges) {
+    addEdge(consumer, child);
+  }
+  return true;
+}
+
+template <typename T>
+bool DynamicDAG<T>::contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer) {
+  // Basically a modified dfsForward
+  vertex_list<T> stack;
+  vertex_list<T> visited;
+  stack.push_back(producer);
+
+  size_t upper_bound = consumer->ord;
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    if (vertex->visited) continue;
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->out_edges) {
+      if (vertex != producer && next->ord == upper_bound) {
+        clearVisited(visited);
+        return true;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
+      }
+    }
+  }
+  clearVisited(visited);
+  return false;
+}
+
+
+// Performs a forward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+// Returns false immediately if upper_bound is reached.
+// Returns true otherwise.
```

Comment:
The return flag is kind of hard to interpret. I think it would be much clearer if we could have an address of a "forbidden vertex" which shouldn't be reachable (the other endpoint of the inserted edge), and throw from here if we see it.

Reply:
Oh I see that's what `upper_bound` is. Can we just move the throw to this function?

Reply:
Yeah dfs is doing a few things for the sake of optimization. upper_bound is both an upper bound and the address of a "forbidden node" -- I think I'll separate the two so the logic is clearer.